extensions/amp-autocomplete/0.1/test-e2e/test-amp-autocomplete.js

```diff
@@ -78,4 +78,23 @@ describes.endtoend('amp-autocomplete', {
     await expect(newItemElements).to.have.length(2);
   });
 
+  it('<amp-autocomplete> should select an item', async() => {
+    const renderedResults =
+      await controller.findElement('.i-amphtml-autocomplete-results');
+    const focusButton = await controller.findElement('#focusButton');
+
+    // Displays all suggested items on focus.
+    await controller.click(focusButton);
+    const itemElements =
+      await controller.findElements('.i-amphtml-autocomplete-item');
+    await expect(controller.getElementAttribute(renderedResults, 'hidden'))
+        .to.be.null;
+    await expect(itemElements).to.have.length(3);
+    await controller.click(itemElements[0]);
+
+    // Displays no items after selecting one.
+    await expect(controller.getElementAttribute(renderedResults, 'hidden'))
+        .not.to.be.null;
```

Comment:
@cvializ Is there a way to convey "controller should not find elements" of a given selector? In this case, I'd want to confirm at the end that `.i-amphtml-autocomplete-item` elements is undefined or the children of the `renderedResults` has length 0.

Reply:
Since `findElement` will wait until it finds the element or times out, you can check the length of a known element's children and expect it to be `0`. e.g.

```js
await expect(controller.getElementProperty(element, 'children')).to.have.length(0);
```

This should also poll until the value becomes `0` if it isn't when it first checks.

Reply:
It looks like for this last test in the shadow-demo environment, clicking on the element `itemElements[0]` fires a `mousedown` event where the `event.target` is a `div.amp-doc-host` element instead of the item element I was expecting. This of course means that "apple" doesn't get selected, the ensuring checks aren't true, and the test fails. Can you explain more what this third environment is doing differently from the first two?

Reply:
There are cases where it can appear that the event is coming from the shadow root, since the browser is unable to expose information about what happens inside the shadow root without explicitly entering it, similar to same-origin iframes. How are you checking where the event comes from?

Reply:
To check where the event comes from, I set a breakpoint on autocomplete's `selectItem_` method, manually mimic the click in the test browser that gets launched by the `gulp e2e` command, and read the local variables in the console.
<img width="401" alt="Screen Shot 2019-04-25 at 4 41 20 PM" src="https://user-images.githubusercontent.com/10456171/56767155-0428d780-6779-11e9-8962-abcbc2bc28c7.png">